src/server/pkg/storage/storage-format-spec.md

```diff
@@ -128,42 +132,135 @@ important differences between these two types of objects as it pertains to GC is
-perform garbage collection we'll need to scan objects and check them
-against our bloom filter to see if they're still referenced. I think this
-can work fine as a process that is always passively running.
-
-The other big difficulty with Pachyderm GC is race conditions. Because of
-our deduplication system, an object that is unreferenced and thus ready to be
-GCed may at any point get re-referenced and thus cease to be ready for GC.
-I think we can solve this with a generational system wherein each object
-belongs to a generation, and we record a bloom filter that describes the
-reference changes in a generation. By summing all the generations together
-we can get a global bloom filter for a generation and figure out what's
-safe to delete. This is best demonstrated with an example:
-
-| gen | ops          | bloom     | bloom sum |
-| --- | ------------ | ------    | --------- |
-|   0 | +foo         | 100101    | 100101    |
-|   1 | +bar         | 001011    | 101112    |
-|   2 | -foo         | -100-10-1 | 001011    |
-
-After generation 2, foo is ready to be deleted, which is reflected in the
-fact that the bloom sum returns negative for foo's bloom filter (100101).
-The tricky part is that if we now want to re-add foo in gen 3 we'll need
-to make sure that the GC process for gen 2 has either already finished,
-because if it hasn't, it might GC foo after we write it for gen 3, or
-knows not to GC this object, because it's about to be re-referenced.
-I think all of this is manageable, but it's definitely tricky.
+### Requirements
+
+* Always-running background process/service
+* No global locks or stoppages required during steady-state operations
+* Iterative and rate-limitable for any large operations
+* Batches updates from Pachyderm and batches deletes to object storage
+* Recovers smoothly from disorderly shutdown
+* Capable of becoming shardable with minimal changes
+
+### API
+
+The garbage collector will have two components, a client-side library and a
+service that runs in pachd.  The library will act as the single-point of contact
+for a client, while coordinating between the pachd service and the database as
+necessary.
+
+```go
+func (gcc *GcClient) ReserveChunks(chunks []Chunk)
+func (gcc *GcClient) UpdateReferences(add []Chunk, remove []Chunk)
+```
+
+The garbage collection service will provide a GRPC API for use by the client
+library (which should not be used directly).
+
+```
+rpc FlushDeletes(FlushDeletesRequest)
+```
+
+When deleting a chunk, the garbage collection service will need to evaluate the
+transitive references from that chunk to others.  Another service in the storage
+layer must provide an interface to fetch this information.
+
+```
+rpc GetReferences(GetReferencesRequest) stream Chunk
+
+message GetReferencesRequest {
+  Chunk chunk
+}
+```
+
+### Implementation
+
+**Client Library**
+
+`ReserveChunks` will store the set of chunks in new rows in the database with a
```

Comment:
Clients block on this when they want to update a chunk that is currently reserved or being deleted? How does a client communicate that it is done with a set of chunks? Through `UpdateReferences`? If you could elaborate more on what the usage pattern would look like for a client that is uploading chunks, that would be helpful.

Reply:
> Clients block on this when they want to update a chunk that is currently reserved or being deleted?

Yes

> How does a client communicate that it is done with a set of chunks?

Currently this is just a TTL reservation period, which is necessary to be durable against client failure.  Other than that, a client is done with a set of chunks once their references are non-zero.

>  If you could elaborate more on what the usage pattern would look like for a client that is uploading chunks, that would be helpful.

I admit I'm not completely clear on the client use cases, but I was thinking client usage would be something along the lines of:

New data:
1. Generate chunk data and hash keys
2. `ReserveChunks` - Reserve references by hash key
3. Upload chunks to object storage
4. `UpdateReferences` - Add a reference for each chunk used

Merging indexes:
1. Generate chunk data and hash keys
2. `ReserveChunks` - Reserve references by hash key
3. Upload chunks to object storage
4. `UpdateReferences` - Add references to the new chunks, remove references from the old chunks

Reply:
> Currently this is just a TTL reservation period, which is necessary to be durable against client failure. Other than that, a client is done with a set of chunks once their references are non-zero.

So, `UpdateReferences` signals that the client is done with a certain set of chunks?

Also, to be clear, is this the pattern for an upload?
```
ReserveChunks
If object does not exist:
  Upload it
UpdateReferences
```

Reply:
Yes, although since it's done in batches, it'd be a bit more like:
```
ReserveChunks
For each chunk:
  If object does not exist:
    Upload it
UpdateReferences
```